snorkel/contrib/learning/relstm.py

```diff
@@ -0,0 +1,231 @@
+import cPickle
+import numpy as np
+import tensorflow as tf
+
+from snorkel.learning import get_train_idxs, TFNoiseAwareModel
+from time import time
+
+
+class reLSTM(TFNoiseAwareModel):
+
+    def __init__(self, save_file=None, name='reLSTM'):
+        """LSTM for relation extraction"""
+        # Define metadata
+        self.mx_len          = None
+        self.word_dict       = SymbolTable()
+        self.dim             = None
+        self.n_v             = None
+        self.lr              = None
+        self.dropout         = None
+        # Define input layers
+        self.sentences       = None
+        self.sentence_length = None
+        self.y               = None
+        # Super constructor
+        super(reLSTM, self).__init__(save_file=save_file, name=name)
+
+    def _mark(self, l, h, idx):
+        """Produce markers based on argument positions"""
+        return [(l, "{}{}".format('[[', idx)), (h+1, "{}{}".format(idx, ']]'))]
+
+    def _mark_sentence(self, s, mids):
+        """Insert markers around relation arguments in word sequence
+        Example: Then Barack married Michelle.  
+             ->  Then [[0 Barack 0]] married [[1 Michelle 1]].
+        """
+        marks = sorted([y for m in mids for y in self._mark(*m)], reverse=True)
+        x = list(s)
+        for k, v in marks:
+            x.insert(k, v)
+        return x
+
+    def _preprocess_data(self, candidates, extend=False):
+        """Convert candidate sentences to lookup sequences"""
+        sentences = []
+        for c in candidates:
+            # Get arguments and lemma sequence
+            args = [
+                (c[0].get_word_start(), c[0].get_word_end(), 1),
+                (c[1].get_word_start(), c[1].get_word_end(), 2)
+            ]
+            s = self._mark_sentence(
+                [w.lower() for w in c.get_parent().lemmas], args
+            )
+            # Either extend word table or retrieve from it
+            retriever = self.word_dict.get if extend else self.word_dict.lookup
+            sentences.append(np.array([retriever(w) for w in s]))
+        return sentences
+
+    def _make_tensor(self, x):
+        """Construct input tensor with padding"""
+        batch_size = len(x)
+        tx = np.zeros((self.mx_len, batch_size), dtype=np.int32)
+        tlen = np.zeros(batch_size, dtype=np.int32)
+        # Pad or trim each x
+        # TODO: fix for arguments outside max length
```

Comment:
Either fix or explain what's going on here / what is currently implemented pre-fix